README.md
# tinycolor2
> TinyColor is a small library for Flutter color manipulation and conversion

A port of [tinycolor2](https://github.com/bgrins/TinyColor) by [Brian Grinstead](https://github.com/bgrins)

## Getting Started

A `TinyColor` receives a `Color` as parameter.

```dart
import 'package:tinycolor2/tinycolor2.dart';

final TinyColor tinyColor = TinyColor.fromColor(Colors.green);
```
Now you can also use the package to extend the native `Color` class with all the same features, but simpler.

## Constructors

### From Flutter's `Color`

```dart
TinyColor.fromColor(Colors.blue);
```

### From a Hex String

The package uses [Pigment](https://pub.dartlang.org/packages/pigment) by [Bregy Malpartida Ramos](https://github.com/bregydoc/) to convert strings to `Color`

```dart
TinyColor.fromString('#FE5567');
```

### From `HSLColor`

```dart
final HSLColor color = HSLColor.fromAHSL(1.0, 250, 57, 30);
TinyColor.fromHSLColor(color);
```

### From `HSVColor`

```dart
final HSVColor color = HSVColor.fromAHSV(1.0, 250, 57, 30);
TinyColor.fromHSVColor(color);
```

## Methods

### Conversions

#### `.toColor()`

Returns the Flutter's `Color`

```dart
final Color color = TinyColor.fromColor(Colors.white).toColor();
```

#### `.toHSLColor()`

Returns the Flutter's `HSLColor`

```dart
final HSLColor color = TinyColor.fromColor(Colors.white).toHSLColor();
```

#### `.toHSVColor()`

Returns the Flutter's `HSVColor`

```dart
final HSVColor color = TinyColor.fromColor(Colors.white).toHSVColor();
```

### Properties

#### `.isLight()`

Return a boolean indicating whether the color's perceived brightness is light.

```dart
TinyColor.fromString("#ffffff").isLight(); // true
TinyColor.fromString("#000000").isLight(); // false
// or with Color extension
Colors.white.isLight;  // true
```

#### `.isDark()`

Return a boolean indicating whether the color's perceived brightness is dark.

```dart
TinyColor.fromString("#ffffff").isDark(); // false
TinyColor.fromString("#000000").isDark(); // true
// or with Color extension
Colors.white.isDark;  // false
```

#### `.getBrightness()`

Returns the perceived brightness of a color, from `0` to `255`, as defined by [Web Content Accessibility Guidelines (Version 1.0)](https://www.w3.org/TR/AERT#color-contrast).

```dart
TinyColor.fromString("#ffffff").getBrightness(); // 255
TinyColor.fromString("#000000").getBrightness(); // 0
// or with Color extension
Colors.grey.brightness;  // 127
```

#### `.getLuminance()`

Return the perceived luminance of a color, a shorthand for flutter `Color.computeLuminance()`

```dart
TinyColor.fromString("#ffffff").getLuminance();
// or with Color extension
Colors.white.luminance;
```

### Utilities

#### `.clone()`

`clone: function() -> TinyColor`. Instantiate a new `TinyColor` object with the same color. Any changes to the new one won't affect the old one.

```dart
final TinyColor color1 = TinyColor.fromColor(Colors.red);
final TinyColor color2 = color1.clone();
color2.setAlpha(20);
```

#### `==`

`==: function(Object) -> bool`. Compares if `[Object]` is the same `TinyColor` object.

```dart
final TinyColor color1 = TinyColor.fromColor(Colors.blue);
final TinyColor color2 = TinyColor.fromColor(Colors.yellow);
if (color1 == color2) return "same";
else return "different";
```

### Modifications

These methods manipulate the current color, and return it for chaining.

```dart
TinyColor.fromColor(Colors.red).lighten().desaturate().color;
// or with Color extension
Colors.red.lighten().desaturate();
```

#### `.setAlpha()`

Sets the alpha value on the current color, from `0` to `255`.

```dart
TinyColor.fromColor(Colors.red).setAlpha(10);
```

#### `.setOpacity()`

Sets the opacity value on the current color, from `0.0` to `1.0`.

```dart
TinyColor.fromColor(Colors.red).setOpacity(0.5);
```

#### `.lighten()`

`lighten: function(amount = 10) -> TinyColor`. Lighten the color a given amount, from `0` to `100`. Providing `100` will always return white.

```dart
TinyColor.fromColor(Colors.red).lighten().color;
TinyColor.fromColor(Colors.red).lighten(100).color;
// or with Color extension
Colors.red.lighten(50);
```

#### `.brighten()`

`brighten: function(amount = 10) -> TinyColor`. Brighten the color a given amount, from `0` to `100`.

```dart
TinyColor.fromColor(Colors.black).brighten().color;
// or with Color extension
Colors.black.brighten(50);
```

#### `.darken()`

`darken: function(amount = 10) -> TinyColor`. Darken the color a given amount, from `0` to `100`. Providing `100` will always return black.

```dart
TinyColor.fromColor(Colors.red).darken().color;
TinyColor.fromColor(Colors.red).darken(100).color;
// or with Color extension
Colors.red.darken(50);
```

#### `.tint()`

Mix the color with pure white, from `0` to `100`. Providing `0` will do nothing, providing `100` will always return white.

```dart
TinyColor.fromColor(Color.red).tint().color;
TinyColor.fromColor(Color.red).tint(100).color;
// or with Color extension
Colors.red.tint(50);
```

#### `.shade()`

Mix the color with pure black, from `0` to `100`. Providing `0` will do nothing, providing `100` will always return black.

```dart
TinyColor.fromColor(Colors.red).shade().color;
TinyColor.fromColor(Colors.red).shade(100).color;
// or with Color extension
Colors.red.shade(50);
```

#### `.desaturate()`

`desaturate: function(amount = 10) -> TinyColor`. Desaturate the color a given amount, from `0` to `100`. Providing `100` will is the same as calling `greyscale`.

```dart
TinyColor.fromColor(Colors.red).desaturate().color;
TinyColor.fromColor(Colors.red).desaturate(100).color;
// or with Color extension
Colors.red.desaturate(50);
```

#### `.saturate()`

`saturate: function(amount = 10) -> TinyColor`. Saturate the color a given amount, from `0` to `100`.

```dart
TinyColor.fromColor(Colors.red).saturate().color;
// or with Color extension
Colors.red.saturate(50);
```

#### `.greyscale()`

`greyscale: function() -> TinyColor`. Completely desaturates a color into greyscale. Same as calling `desaturate(100)`.

```dart
TinyColor.fromColor(Colors.red).greyscale().color;
// or with Color extension
Colors.red.greyscale();
```

#### `.spin()`

`spin: function(amount) -> TinyColor`. Spin the hue a given amount, from `-360` to `360`. Calling with `0`, `360`, or `-360` will do nothing (since it sets the hue back to what it was before).

```dart
TinyColor.fromColor(Colors.red).spin(180).color;
// or with Color extension
Colors.red.spin(180);

// spin(0) and spin(360) do nothing
TinyColor.fromColor(Colors.red).spin(0).color;
TinyColor.fromColor(Colors.red).spin(360).color;
```

#### `.complement()`

`complement: function() -> TinyColor`. Returns the complementary color for dynamic matching.

```dart
TinyColor.fromColor(Colors.red).complement().color;
// or with Color extension
Colors.red.complement();
```

#### `.mix()`

`mix: function(toColor, amount = 50) -> TinyColor`. Blends the color with another color a given amount, from `0` to `100`.

```dart
TinyColor.fromColor(Colors.red).mix(TinyColor.fromColor(Colors.yellow, 20)).color;
// or with Color extension
Colors.red.mix(Colors.yellow, 20);
```

## Extensions

All methods listed above are included in extensions. Note that some methods has changed to return a `bool` or `double` value instead to match with Dart approach, and the rest would return its original class type instead of `TinyColor`.

### `Color`

```dart
final TinyColor color = Colors.black.toTinyColor();
final HSVColor hsv = Colors.white.toHSVColor();
final HSLColor hsl = Colors.white.toHSLColor();
final bool isDark = Colors.yellow.isDark;
final bool isLight = Colors.red.isLight;
final double luminance = Colors.blue.luminance;
final double brightness = Colors.blue.brightness;
final Color newColor = Colors.green.mix(Colors.white, 50);
```


CHANGELOG.md
# Changelog

## [3.0.1] - 2022-09-19

### Fixed
- Correct pubspec Flutter + Dart SDK versions at [PR #64](https://github.com/TinyCommunity/tinycolor2/pull/64)
- Remove upper bound in pubspec Flutter + Dart SDK versions at [PR #65](https://github.com/TinyCommunity/tinycolor2/pull/65)

## [3.0.0] - 2022-09-12

### Changed
- Upgrade to Flutter 3.3.0 + Upgrade lint package to 1.10.0 at [PR #63](https://github.com/TinyCommunity/tinycolor2/pull/63)

## [2.1.0] - 2022-06-20

### Changed
* Rename constructors and methods per Dart & Flutter naming conventions at [PR #41](https://github.com/TinyCommunity/tinycolor2/pull/41)
* `HslColor` class is now deprecated in favor of native `HSLColor` at [PR #18](https://github.com/TinyCommunity/tinycolor2/pull/35)
* Deprecated TinyColor.fromRGB() at [PR #40](https://github.com/TinyCommunity/tinycolor2/pull/40)
* Refactor and improve tests at [PR #57](https://github.com/TinyCommunity/tinycolor2/pull/57)
* Update extension codes to remove usage of deprecated methods and add new methods at [PR #59](https://github.com/TinyCommunity/tinycolor2/pull/59)

### Fixed
* Not validating range of parameters at [PR #55](https://github.com/TinyCommunity/tinycolor2/pull/55)
* Split up test for parameterized tests at [PR #58](https://github.com/TinyCommunity/tinycolor2/pull/58)
* Most setters should update internal color value and return this at [PR #42](https://github.com/TinyCommunity/tinycolor2/pull/42)

## [2.0.1] - 2022-03-11

### Changed
* Rewrite tests at [PR #35](https://github.com/TinyCommunity/tinycolor2/pull/35)

### Fixed
* `.mix()` not matching readme usage at [PR #44](https://github.com/TinyCommunity/tinycolor2/pull/44)
* `.greyscale()` should be a method, not a property in `Color` extension at [PR #45](https://github.com/TinyCommunity/tinycolor2/pull/45)
* Typo in `.complement()` naming at [PR #43](https://github.com/TinyCommunity/tinycolor2/pull/43)
* `alpha` range should be `0`-`255` at [PR #39](https://github.com/TinyCommunity/tinycolor2/pull/39)

## [2.0.0] - 2021-07-28

### Added
* `.toHex8()` at [PR #17](https://github.com/TinyCommunity/tinycolor2/pull/17)
* `.equals()` and `operator ==` at [PR #16](https://github.com/TinyCommunity/tinycolor2/pull/16)
* Example project at [PR #13](https://github.com/TinyCommunity/tinycolor2/pull/13)

### Changed
* Forked as a community version with new name `tinycolor2`
* Null safety migration w/ dependency upgrade at [PR #24](https://github.com/TinyCommunity/tinycolor2/pull/24)
* Library restructure at [PR #5](https://github.com/TinyCommunity/tinycolor2/pull/5)

### Fixed
* #2, #6, #10 from upstream at [PR #5](https://github.com/TinyCommunity/tinycolor2/pull/5)

## [1.0.3] - 2020-08-27

### Added
* `Color` class extension method for direct usage

## [1.0.2] - 2018-08-17

### Fixed
* Compatibility issues with Dart `2.1` and `pigment` `1.0.3`

## [1.0.1] - 2018-08-04

### Fixed
* `HSLColor` class name duplicate definition

### Changed
* Package description

## [1.0.0] - 2018-08-04

### Added
* Initial port of `tinycolor2`


test/tinycolor_test.dart
import 'package:flutter/painting.dart' show Color, HSLColor, HSVColor;
import 'package:flutter_test/flutter_test.dart';
import 'package:tinycolor2/tinycolor2.dart';

void main() {
  group(
    "TinyColor.fromColor()",
    () {
      TinyColor? color;
      setUp(() {
        color = TinyColor.fromColor(const Color(0xFFFFFFFF));
      });
      tearDown(() {
        color = null;
      });
      test(
        ".toColor()",
        () {
          expect(
            color!.toColor(),
            const Color(0xFFFFFFFF),
          );
        },
      );
      test(
        ".toHex8()",
        () {
          expect(
            color!.toHex8(),
            "ffffffff",
          );
        },
      );
      test(
        ".getBrightness()",
        () {
          expect(
            color!.getBrightness(),
            255,
          );
        },
      );
      test(
        ".isLight()",
        () {
          expect(
            color!.isLight(),
            true,
          );
        },
      );
      test(
        ".isDark()",
        () {
          expect(
            color!.isDark(),
            false,
          );
        },
      );
      test(
        ".getLuminance()",
        () {
          expect(
            color!.getLuminance(),
            const Color(0xFFFFFFFF).computeLuminance(),
          );
        },
      );
      test(
        ".setAlpha()",
        () {
          expect(
            color!.setAlpha(0x00).color.alpha,
            0x00,
          );
        },
      );
      test(
        ".setOpacity()",
        () {
          // underlying dart implementation converts the opacity value to an
          // int, then back into a double. Thus some precision is loss.
          expect(
            color!.setOpacity(0.5).color.opacity,
            moreOrLessEquals(0.5, epsilon: 1e-2),
          );
        },
      );
      group(
        ".lighten()",
        () {
          test(
            "0",
            () {
              expect(
                color!.lighten(0).color,
                const Color(0xFFFFFFFF),
              );
            },
          );
          test(
            "100",
            () {
              expect(
                color!.lighten(100).color,
                const Color(0xFFFFFFFF),
              );
            },
          );
          test(
            "default",
            () {
              expect(
                color!.lighten().color,
                const Color(0xFFFFFFFF),
              );
            },
          );
        },
      );
      group(
        ".brighten()",
        () {
          test(
            "0",
            () {
              expect(
                color!.brighten(0).color,
                const Color(0xFFFFFFFF),
              );
            },
          );
          test(
            "100",
            () {
              expect(
                color!.brighten(100).color,
                const Color(0xFFFFFFFF),
              );
            },
          );
          test(
            "default",
            () {
              expect(
                color!.brighten().color,
                const Color(0xFFFFFFFF),
              );
            },
          );
        },
      );
      group(
        ".darken()",
        () {
          test(
            "0",
            () {
              expect(
                color!.darken(0).color,
                const Color(0xFFFFFFFF),
              );
            },
          );
          test(
            "100",
            () {
              expect(
                color!.darken(100).color,
                const Color(0xFF000000),
              );
            },
          );
          test(
            "default",
            () {
              expect(
                color!.darken().color,
                const Color(0xFFE6E6E6),
              );
            },
          );
        },
      );
      group(
        ".tint()",
        () {
          test(
            "0",
            () {
              expect(
                color!.tint(0).color,
                const Color(0xFFFFFFFF),
              );
            },
          );
          test(
            "100",
            () {
              expect(
                color!.tint(100).color,
                const Color(0xFFFFFFFF),
              );
            },
          );
          test(
            "default",
            () {
              expect(
                color!.tint().color,
                const Color(0xFFFFFFFF),
              );
            },
          );
        },
      );
      group(
        ".shade()",
        () {
          test(
            "0",
            () {
              expect(
                color!.shade(0).color,
                const Color(0xFFFFFFFF),
              );
            },
          );
          test(
            "100",
            () {
              expect(
                color!.shade(100).color,
                const Color(0xFF000000),
              );
            },
          );
          test(
            "default",
            () {
              expect(
                color!.shade().color,
                const Color(0xFFE6E6E6),
              );
            },
          );
        },
      );
      group(
        ".desaturate()",
        () {
          test(
            "0",
            () {
              expect(
                color!.desaturate(0).color,
                const Color(0xFFFFFFFF),
              );
            },
          );
          test(
            "100",
            () {
              expect(
                color!.desaturate(100).color,
                const Color(0xFFFFFFFF),
              );
            },
          );
          test(
            "default",
            () {
              expect(
                color!.desaturate().color,
                const Color(0xFFFFFFFF),
              );
            },
          );
        },
      );
      group(
        ".saturate()",
        () {
          test(
            "0",
            () {
              expect(
                color!.saturate(0).color,
                const Color(0xFFFFFFFF),
              );
            },
          );
          test(
            "100",
            () {
              expect(
                color!.saturate(100).color,
                const Color(0xFFFFFFFF),
              );
            },
          );
          test(
            "default",
            () {
              expect(
                color!.saturate().color,
                const Color(0xFFFFFFFF),
              );
            },
          );
        },
      );
      test(
        ".greyscale()",
        () {
          expect(
            color!.greyscale().color,
            const Color(0xFFFFFFFF),
          );
        },
      );
      group(
        ".spin()",
        () {
          test(
            "-360",
            () {
              expect(
                color!.spin(-360).color,
                const Color(0xFFFFFFFF),
              );
            },
          );
          test(
            "-180",
            () {
              expect(
                color!.spin(-180).color,
                const Color(0xFFFFFFFF),
              );
            },
          );
          test(
            "0",
            () {
              expect(
                color!.spin(0).color,
                const Color(0xFFFFFFFF),
              );
            },
          );
          test(
            "180",
            () {
              expect(
                color!.spin(180).color,
                const Color(0xFFFFFFFF),
              );
            },
          );
          test(
            "360",
            () {
              expect(
                color!.spin(360).color,
                const Color(0xFFFFFFFF),
              );
            },
          );
        },
      );
      test(
        ".complement()",
        () {
          expect(
            color!.complement().color,
            const Color(0xFFFFFFFF),
          );
        },
      );
      group(
        ".mix()",
        () {
          test(
            "0",
            () {
              expect(
                color!.mix(const Color(0xFF000000), 0).color,
                const Color(0xFFFFFFFF),
              );
            },
          );
          test(
            "100",
            () {
              expect(
                color!.mix(const Color(0xFF000000), 100).color,
                const Color(0xFF000000),
              );
            },
          );
          test(
            "default",
            () {
              expect(
                color!.mix(const Color(0xFF000000)).color,
                const Color(0xFF808080),
              );
            },
          );
        },
      );
      test(
        ".clone()",
        () {
          expect(
            color!.clone(),
            TinyColor.fromColor(const Color(0xFFFFFFFF)),
          );
        },
      );
      test(
        "==",
        () {
          expect(
            color,
            TinyColor.fromColor(const Color(0xFFFFFFFF)),
          );
        },
      );
    },
  );

  group(
    "TinyColor.fromHSLColor()",
    () {
      final TinyColor color =
          TinyColor.fromHSLColor(HSLColor.fromColor(const Color(0xFFFFFFFF)));
      test(
        ".toColor()",
        () {
          expect(
            color.toColor(),
            const Color(0xFFFFFFFF),
          );
        },
      );
      test(
        ".toHSLColor()",
        () {
          expect(
            color.toHSLColor(),
            HSLColor.fromColor(const Color(0xFFFFFFFF)),
          );
        },
      );
    },
  );

  group(
    "TinyColor.fromHSVColor()",
    () {
      final TinyColor color =
          TinyColor.fromHSVColor(HSVColor.fromColor(const Color(0xFFFFFFFF)));
      test(
        ".toColor()",
        () {
          expect(
            color.toColor(),
            const Color(0xFFFFFFFF),
          );
        },
      );
      test(
        ".toHSVColor()",
        () {
          expect(
            color.toHSVColor(),
            HSVColor.fromColor(const Color(0xFFFFFFFF)),
          );
        },
      );
    },
  );
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:tinycolor2/tinycolor2.dart';

void main() {
  runApp(TinyColorApp());
}

class TinyColorApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'TinyColor2 Example',
      theme: ThemeData(
        primarySwatch: Colors.grey,
      ),
      home: const ExamplePage(),
    );
  }
}

class ExamplePage extends StatelessWidget {
  const ExamplePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final TinyColor tc = TinyColor.fromColor(Colors.blue);

    return Scaffold(
      appBar: AppBar(
        title: const Text("TinyColor2 Example"),
      ),
      body: ListView(
        children: <Widget>[
          _createListItem(
            title: "Original color",
            subtitle: """
TinyColor tc = TinyColor(Colors.blue)
Color c = Colors.blue
""",
            color: tc.color,
            showOr: false,
          ),
          _createListItem(
            title: "tc.lighten(20)",
            subtitle: "c.lighten(20)",
            color: tc.lighten(20).color,
          ),
          _createListItem(
            title: "tc.brighten(20)",
            subtitle: "c.brighten(20)",
            color: tc.brighten(20).color,
          ),
          _createListItem(
            title: "tc.darken(20)",
            subtitle: "c.darken(20)",
            color: tc.darken(20).color,
          ),
          _createListItem(
            title: "tc.tint(20)",
            subtitle: "c.tint(20)",
            color: tc.tint(20).color,
          ),
          _createListItem(
            title: "tc.shade(20)",
            subtitle: "c.shade(20)",
            color: tc.shade(20).color,
          ),
          _createListItem(
            title: "tc.desaturate(20)",
            subtitle: "c.desaturate(20)",
            color: tc.desaturate(20).color,
          ),
          _createListItem(
            title: "tc.saturate(20)",
            subtitle: "c.saturate(20)",
            color: tc.saturate(20).color,
          ),
          _createListItem(
            title: "tc.greyscale()",
            subtitle: "c.greyscale()",
            color: tc.greyscale().color,
          ),
          _createListItem(
            title: "tc.spin(90)",
            subtitle: "c.spin(90)",
            color: tc.spin(90).color,
          ),
          _createListItem(
            title: "tc.complement()",
            subtitle: "c.complement()",
            color: tc.complement().color,
          ),
          _createListItem(
            title: "tc.mix(Colors.yellow, 20)",
            subtitle: "c.mix(Colors.yellow, 20)",
            color: tc.mix(Colors.yellow, 20).color,
          ),
        ],
      ),
    );
  }

  Widget _createListItem({
    required String title,
    required String subtitle,
    required Color color,
    bool showOr = true,
  }) =>
      ListTile(
        title: Text(title),
        subtitle: Text("${showOr ? "or: " : ""}$subtitle"),
        trailing: CircleAvatar(backgroundColor: color),
      );
}


lib/src/conversion.dart
import 'dart:math' as math;

import 'package:flutter/painting.dart' show Color, HSLColor, HSVColor;
import 'package:tinycolor2/src/util.dart';

HSLColor rgbToHsl({
  required double r,
  required double g,
  required double b,
}) {
  final double rBound = bound01(r, 255.0);
  final double gBound = bound01(g, 255.0);
  final double bBound = bound01(b, 255.0);

  final max = [rBound, gBound, bBound].reduce(math.max);
  final min = [rBound, gBound, bBound].reduce(math.min);
  double h = 0.0;
  double s = 0.0;
  final double l = (max + min) / 2;

  if (max == min) {
    h = s = 0.0;
  } else {
    final double d = max - min;
    s = l > 0.5 ? d / (2.0 - max - min) : d / (max + min);
    if (max == rBound) {
      h = (gBound - bBound) / d + (gBound < bBound ? 6 : 0);
    } else if (max == gBound) {
      h = (bBound - rBound) / d + 2;
    } else if (max == bBound) {
      h = (rBound - gBound) / d + 4;
    }
  }

  h /= 6.0;
  return HSLColor.fromAHSL(1.0, h, s, l);
}

Color hslToRgb(HSLColor hsl) {
  double r;
  double g;
  double b;
  final double h = bound01(hsl.hue, 360.0);
  final double s = bound01(hsl.saturation * 100, 100.0);
  final double l = bound01(hsl.lightness * 100, 100.0);

  if (s == 0.0) {
    r = g = b = l;
  } else {
    final q = l < 0.5 ? l * (1.0 + s) : l + s - l * s;
    final p = 2 * l - q;
    r = _hue2rgb(p, q, h + 1 / 3);
    g = _hue2rgb(p, q, h);
    b = _hue2rgb(p, q, h - 1 / 3);
  }
  return Color.fromARGB(
    hsl.alpha.round(),
    (r * 255).round(),
    (g * 255).round(),
    (b * 255).round(),
  );
}

HSVColor rgbToHsv({
  required int r,
  required int g,
  required int b,
  required int a,
}) =>
    HSVColor.fromColor(Color.fromARGB(a, r, g, b));

double _hue2rgb(double p, double q, double t) {
  double tint = t;

  if (tint < 0) tint += 1;
  if (tint > 1) tint -= 1;
  if (tint < 1 / 6) return p + (q - p) * 6 * tint;
  if (tint < 1 / 2) return q;
  if (tint < 2 / 3) return p + (q - p) * (2 / 3 - tint) * 6;
  return p;
}


lib/src/extensions.dart
import 'package:flutter/painting.dart';
import 'package:tinycolor2/src/tinycolor.dart';

/// Extends the Color class to allow direct TinyColor manipulation natively
extension TinyColorExtension on Color {
  /// Converts standard Color to TinyColor object
  TinyColor toTinyColor() => TinyColor.fromColor(this);

  HSVColor toHSVColor() => TinyColor.fromColor(this).toHSVColor();

  HSLColor toHSLColor() => TinyColor.fromColor(this).toHSLColor();

  String toHex8() => TinyColor.fromColor(this).toHex8();

  /// Lighten the color a given amount, from 0 to 100. Providing 100 will always return white.
  Color lighten([int amount = 10]) =>
      TinyColor.fromColor(this).lighten(amount).toColor();

  /// Brighten the color a given amount, from 0 to 100.
  Color brighten([int amount = 10]) =>
      TinyColor.fromColor(this).brighten(amount).toColor();

  /// Darken the color a given amount, from 0 to 100. Providing 100 will always return black.
  Color darken([int amount = 10]) =>
      TinyColor.fromColor(this).darken(amount).toColor();

  /// Mix the color with pure white, from 0 to 100. Providing 0 will do nothing, providing 100 will always return white.
  Color tint([int amount = 10]) =>
      TinyColor.fromColor(this).tint(amount).toColor();

  /// Mix the color with pure black, from 0 to 100. Providing 0 will do nothing, providing 100 will always return black.
  Color shade([int amount = 10]) =>
      TinyColor.fromColor(this).shade(amount).toColor();

  /// Desaturate the color a given amount, from 0 to 100. Providing 100 will is the same as calling greyscale.
  Color desaturate([int amount = 10]) =>
      TinyColor.fromColor(this).desaturate(amount).toColor();

  /// Saturate the color a given amount, from 0 to 100.
  Color saturate([int amount = 10]) =>
      TinyColor.fromColor(this).saturate(amount).toColor();

  /// Completely desaturates a color into greyscale. Same as calling desaturate(100).
  Color greyscale() => TinyColor.fromColor(this).greyscale().toColor();

  /// Spin the hue a given amount, from -360 to 360. Calling with 0, 360, or -360 will do nothing (since it sets the hue back to what it was before).
  Color spin([double amount = 0]) =>
      TinyColor.fromColor(this).spin(amount).toColor();

  /// Blends the color with another color a given amount, from 0 - 100, default 50.
  Color mix(Color toColor, [int amount = 50]) =>
      TinyColor.fromColor(this).mix(toColor, amount).toColor();

  /// Returns the complementary color for dynamic matching
  Color complement() => TinyColor.fromColor(this).complement().toColor();

  /// Returns the perceived brightness of a color, from 0-255, as defined by Web Content Accessibility Guidelines (Version 1.0).Returns the perceived brightness of a color, from 0-255, as defined by Web Content Accessibility Guidelines (Version 1.0).
  double get brightness => TinyColor.fromColor(this).getBrightness();

  /// Return the perceived luminance of a color, a shorthand for flutter Color.computeLuminance
  double get luminance => TinyColor.fromColor(this).getLuminance();

  /// Return a boolean indicating whether the color's perceived brightness is light.
  bool get isLight => TinyColor.fromColor(this).isLight();

  /// Return a boolean indicating whether the color's perceived brightness is dark.
  bool get isDark => TinyColor.fromColor(this).isDark();
}


lib/src/tinycolor.dart
import 'dart:math' as math;

import 'package:flutter/painting.dart' show Color, HSLColor, HSVColor;
import 'package:pigment/pigment.dart';
import 'package:tinycolor2/src/util.dart';

class TinyColor {
  final Color originalColor;
  Color _color;

  @Deprecated('Use TinyColor.fromColor() instead.')
  TinyColor(this.originalColor) : _color = Color(originalColor.value);

  TinyColor.fromColor(this.originalColor) : _color = Color(originalColor.value);

  @Deprecated('Use TinyColor.fromColor(Color.fromARGB(a, r, g, b)) instead.')
  factory TinyColor.fromRGB({
    required int r,
    required int g,
    required int b,
    int a = 255,
  }) =>
      TinyColor.fromColor(Color.fromARGB(a, r, g, b));

  @Deprecated('Use TinyColor.fromHSLColor() instead.')
  factory TinyColor.fromHSL(HSLColor hsl) => TinyColor.fromColor(hsl.toColor());

  factory TinyColor.fromHSLColor(HSLColor hsl) =>
      TinyColor.fromColor(hsl.toColor());

  @Deprecated('Use TinyColor.fromHSVColor() instead.')
  factory TinyColor.fromHSV(HSVColor hsl) => TinyColor.fromHSVColor(hsl);

  factory TinyColor.fromHSVColor(HSVColor hsv) =>
      TinyColor.fromColor(hsv.toColor());

  factory TinyColor.fromString(String string) =>
      TinyColor.fromColor(Pigment.fromString(string));

  bool isDark() => getBrightness() < 128.0;

  bool isLight() => !isDark();

  double getBrightness() =>
      (_color.red * 299 + _color.green * 587 + _color.blue * 114) / 1000;

  double getLuminance() => _color.computeLuminance();

  TinyColor setAlpha(int alpha) {
    _color = _color.withAlpha(alpha);
    return this;
  }

  TinyColor setOpacity(double opacity) {
    _color = _color.withOpacity(opacity);
    return this;
  }

  Color toColor() => Color(_color.value);

  @Deprecated('Use TinyColor.toHSVColor() instead.')
  HSVColor toHsv() => toHSVColor();

  HSVColor toHSVColor() => HSVColor.fromColor(_color);

  @Deprecated('Use TinyColor.toHSLColor() instead.')
  HSLColor toHsl() => HSLColor.fromColor(_color);

  HSLColor toHSLColor() => HSLColor.fromColor(_color);

  String toHex8() => _color.value.toRadixString(16).padLeft(8, '0');

  TinyColor clone() => TinyColor.fromColor(_color);

  TinyColor lighten([int amount = 10]) {
    final p = RangeError.checkValueInInterval(amount, 0, 100, 'amount') / 100;
    final hsl = toHSLColor();
    final lightness = clamp01(hsl.lightness + p);
    _color = hsl.withLightness(lightness).toColor();
    return this;
  }

  TinyColor brighten([int amount = 10]) {
    final p = RangeError.checkValueInInterval(amount, 0, 100, 'amount') / 100;
    _color = Color.fromARGB(
      _color.alpha,
      math.max(
        0,
        math.min(
          255,
          _color.red - (255 * -p).round(),
        ),
      ),
      math.max(
        0,
        math.min(
          255,
          _color.green - (255 * -p).round(),
        ),
      ),
      math.max(
        0,
        math.min(
          255,
          _color.blue - (255 * -p).round(),
        ),
      ),
    );
    return this;
  }

  TinyColor darken([int amount = 10]) {
    final p = RangeError.checkValueInInterval(amount, 0, 100, 'amount') / 100;
    final hsl = toHSLColor();
    final lightness = clamp01(hsl.lightness - p);
    _color = hsl.withLightness(lightness).toColor();
    return this;
  }

  TinyColor tint([int amount = 10]) => mix(
        const Color.fromRGBO(255, 255, 255, 1.0),
        amount,
      );

  TinyColor shade([int amount = 10]) => mix(
        const Color.fromRGBO(0, 0, 0, 1.0),
        amount,
      );

  TinyColor desaturate([int amount = 10]) {
    final p = RangeError.checkValueInInterval(amount, 0, 100, 'amount') / 100;
    final hsl = toHSLColor();
    final saturation = clamp01(hsl.saturation - p);
    _color = hsl.withSaturation(saturation).toColor();
    return this;
  }

  TinyColor saturate([int amount = 10]) {
    final p = RangeError.checkValueInInterval(amount, 0, 100, 'amount') / 100;
    final hsl = toHSLColor();
    final saturation = clamp01(hsl.saturation + p);
    _color = hsl.withSaturation(saturation).toColor();
    return this;
  }

  TinyColor greyscale() => desaturate(100);

  TinyColor spin(double amount) {
    final hsl = toHSLColor();
    final hue = (hsl.hue + amount) % 360;
    _color = hsl.withHue(hue < 0 ? 360 + hue : hue).toColor();
    return this;
  }

  TinyColor mix(Color toColor, [int amount = 50]) {
    final p = RangeError.checkValueInInterval(amount, 0, 100, 'amount') / 100;
    _color = Color.fromARGB(
      ((toColor.alpha - _color.alpha) * p + _color.alpha).round(),
      ((toColor.red - _color.red) * p + _color.red).round(),
      ((toColor.green - _color.green) * p + _color.green).round(),
      ((toColor.blue - _color.blue) * p + _color.blue).round(),
    );
    return this;
  }

  TinyColor complement() {
    final hsl = toHSLColor();
    final hue = (hsl.hue + 180) % 360;
    _color = hsl.withHue(hue).toColor();
    return this;
  }

  Color get color => toColor();

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is TinyColor &&
          runtimeType == other.runtimeType &&
          color == other.color;

  @override
  int get hashCode => color.hashCode;

  @Deprecated('Use == instead.')
  bool equals(Object other) => this == other;
}
